### DISCUSSION TITLE
harfbuzz normalizes?!

### DISCUSSION THREAD

**Discussion by bdenckla** [harfbuzz/harfbuzz#2942] [Q&A] [answered]
I'm new to harfbuzz, but from what I can see, [it seems to normalize](https://github.com/harfbuzz/harfbuzz/blob/master/src/hb-ot-shape-normalize.cc), which I found surprising.

I am used to having to judiciously insert CGJs into Biblical Hebrew strings to make them robust to normalization _if I want them to be usable in certain applications_, notably web browsers.

But I am surprised to find a piece of software below the application level doing normalization.

For instance on Windows, where Uniscribe/DirectWrite not harfbuzz is the shaper, I am used to the luxury of being able to use CGJ-free strings when I know these strings are only intended for use inside MS Word.

Is no application safe from normalization, if the application uses harfbuzz? Or can harfbuzz be told, optionally, to skip normalization?

Perhaps in many language-script pairs, normalization is just about decomposed vs precomposed forms, and re-orderings that preserve semantics and appearance. But in case you are not aware, this is decidedly _not the case_ with Biblical Hebrew. The re-orderings caused by normalization _do not preserve semantics and appearance_.

Any help greatly appreciated... I may have misunderstood the situation.

Finally, to provide a concrete example of a situation that appears hundreds of times in the Hebrew Bible, here's the first 3 lines of output of `hb-shape` run with `--show-text --show-unicode --trace`:

> (לִָֽם)
> <U+05DC,U+05B8,U+05BD,U+05B4,U+05DD>
> trace: start lookup 0	buffer: [lamed=0|qamats=0|hiriq=0|meteg=0|finalmem=4]

For reference, the original input sequence, which we might abbreviate <DC,B8,BD,B4,DD>, corresponds to:

> <lamed,qamats,meteg,hiriq,finalmem>

From the above we can see that normalization has flipped the order of meteg and hiriq, from meteg-then-hiriq to hiriq-then-meteg. This flip changes semantics and appearance.

(Note for nitpickers: above, in giving names to code points, I can afford to play fast and loose with the code point vs glyph distinction since that is not at issue here. E.g. I can afford to refer to "DD" as "finalmem" even though more formally it is HEBREW LETTER FINAL MEM (the name of the Unicode code point).)

**Comment by jfkthame** [comment]
> normalization has flipped the order of meteg and hiriq, from meteg-then-hiriq to hiriq-then-meteg

Yes, that is indeed what normalization will do, by definition. The Unicode character sequences <meteg, hiriq> and <hiriq, meteg> are defined by Unicode to be *canonically equivalent*, meaning that you cannot expect to reliably distinguish between them and processes are free to treat them identically. That's why to "preserve" a distinction between them, you have to introduce CGJ.

This is (in my opinion, at least) an error in how Unicode encoded these characters, but at this point it is enshrined in the standard and will not be changed. And given this fact, the harfbuzz behavior is correct. It's unfortunate, but the only solution is to encode the sequences of text elements <meteg, hiriq> and <hiriq, meteg> with Unicode sequences that are *not* canonically equivalent, and the way to do that is to insert CGJ, which "blocks" the sorting of the diacritics that normalization would otherwise do.

As you've seen, some software does *not* normalize, and some fonts render the two canonically-equivalent sequences with different appearances. But relying on that behavior is inherently fragile and *will* lead to grief in the end.

**Reply by bdenckla** [reply]
Thanks for your quick reply @jfkthame. At the moment I am using harfbuzz to test fonts for Biblical Hebrew, using [fontbakery](https://github.com/googlefonts/fontbakery/). So, if harfbuzz has no way to disable normalization, I cannot fully test these fonts using harfbuzz. Though I agree that in general it is inadvisable for content creators to use fragile strings (strings not robust to normalization), nonetheless I think a good font needs to handle such strings identically to their robust counterparts, and therefore this handling needs to be tested.

Perhaps a partial solution to this problem would be if fontbakery supported alternate shapers, not just harfbuzz. I raise this possibility in [this fontbakery GitHub discussion](https://github.com/googlefonts/fontbakery/discussions/3235). In particular, if fontbakery supported Uniscribe (assuming Uniscribe even provides the proper "hooks"), then I could test various aspects of a font's behavior under Uniscribe, including (but not limited to) Uniscribe's non-normalizing behavior.

And of course same goes for Apple's Core Text shaper: it would be nice if fontbakery supported Core Text, too (again, assuming Core Text even provides the proper "hooks" for fontbakery).

**Reply by bdenckla** [reply]
With great interest, I found the following on the "[Uniscribe integration](https://harfbuzz.github.io/integration-uniscribe.html)" page of the harfbuzz docs:

> one of HarfBuzz's design goals is to accurately reproduce the same output for shaping a given text segment that Uniscribe produces — even to the point of duplicating known shaping bugs or deviations from the specification — so you can be confident that your users' documents with their existing fonts will not be affected adversely by switching to HarfBuzz

It seems to me this design goal is not met, if HarfBuzz provides no way to turn normalization off.

**Reply by bdenckla** [reply]
For completeness I'll note that HarfBuzz does not have the same goals for Core Text as it has for Uniscribe. From the "[Core Text integration](https://harfbuzz.github.io/integration-coretext.html)" page:
> HarfBuzz does not attempt to offer the same drop-in-replacement functionality for Core Text that it strives for with Uniscribe on Windows

**Reply by behdad** [reply]
If a font produces different rendering for two canonically-equivalent Unicode sequences under Uniscribe, that font is NOT Unicode-compliant. We do not support that use-case.

As you noted, you are interested in this for testing only. fontbakery, if built on windows, can choose to give you option to run HarfBuzz with the Uniscribe backend as well. No amount of testing with HarfBuzz's builtin shaper will guarantee that your font works the exact same way on Uniscribe.

**Reply by bdenckla** [reply]
Thanks for your reply @behdad. I am sympathetic to your definition of a Unicode-compliant font. And I can understand why you would not want to support non-compliant fonts. Yet sometimes "de facto" standards arise that contradict official standards. Since the 3 major Biblical Hebrew font families are all non-compliant in the same way, and Uniscribe supports their non-compliance, this sort of leaves HarfBuzz as the odd man out. Being right often means being lonely! Indeed this was the case for almost all of the prophets of the Hebrew Bible, whose words I am struggling to represent in Unicode: no one listened to them!

For reference, those 3 major families are:

- SBL Hebrew (a "family of one")
- Ezra SIL & Ezra SIL SR
- Taamey Frank CLM & other fonts by Yoram Gnat (of blessed memory)

I am going to try to make some improvements to Yoram Gnat's fonts, which may make me their de facto maintainer.

Perhaps I will also make "forks" of his fonts, and make them Unicode compliant, and give them slightly different names than their originals. I don't think such changes could be considered bug fixes because, from a user's perspective, they would break many documents.

These new compliant fonts would, presumably, perform substitutions that implement the relevant parts of Unicode normalization. One of the many benefits of such compliance is, although the font's behavior might surprise users used to the old, non-compliant behavior, they will be surprised "up front" rather than later. For example they will be surprised "up front" when composing a document in MS Word instead of being surprised later when they view this document in a web browser. I.e. the text will look the same in all environments, regardless of whether the environments are normalizing (e.g. a web browser) or non-normalizing (e.g. MS Word).

**Reply by bdenckla** [reply]
@behdad wrote:
> fontbakery, if built on windows, can choose to give you option to run HarfBuzz with the Uniscribe backend as well.

Would this imply building my own HarfBuzz, with the Uniscribe back-end enabled, as mentioned on the [Uniscribe integration](https://harfbuzz.github.io/integration-uniscribe.html) page, quoted below?

> you may notice a reference to an optional uniscribe shaper back-end in the the section called “Configuration options” section of the HarfBuzz manual. ... It is a internal code path used in the hb-shape command-line utility, which hands shaping functionality over to Uniscribe entirely, when run on a Windows system. That allows testing HarfBuzz's native output against the Uniscribe engine, for tracking compatibility and debugging. Because this back-end is only used when testing HarfBuzz functionality, it is disabled by default when building the HarfBuzz binaries.

**Comment by khaledhosny** [comment]
Hebrew probably can benefit from the equivalent of [UTR #‍53](http://unicode.org/reports/tr53/) which would fix the broken Unicode combining classes while maintaining identical rendering of canonically equivalent strings. HarfBuzz implements UTR #‍53 and I think would be open to implementing its equivalents in other scripts.

**Reply by bdenckla** [reply]
Thanks @khaledhosny for bringing [UTR #53](http://unicode.org/reports/tr53/) to my attention. Indeed it seems that both Arabic and Hebrew combining classes were defined without sufficient input from relevant experts. It is an intriguing option to make Unicode Hebrew "do the right thing" in more cases without use of CGJ. But it seems that again Arabic and Hebrew are similar in that there will still remain cases where CGJ is needed, and therefore the same questions will remain about what the shaper and font should do in cases where CGJ is needed, but omitted. A fix like [UTR #53](http://unicode.org/reports/tr53/) would make those questions less important, which would be nice.

**Reply by alerque** [reply]
> and therefore the same questions will remain about what the shapper and font should do in cases where CGJ is needed, but omitted

I presume "needed" here refers to being prescribed by Unicode. Encoding a string in UTF-8 two different ways that that are defined canonically equivalent and expecting the same result is empirically wrong, hence it is "needed" to disambiguate them with CGJ (or whatever it is).

Taking that definition of needed, if what is "needed" is "omitted" then the fault is with the input string not with the shapper. Trying to second guess **broken** input to "do the right thing" when the spec defines something different is a pathway to madness. Okay anybody involved in Unicod is already half mad ... to even crazier madness.

**Reply by bdenckla** [reply]
@alerque wrote:
> Encoding a string in UTF-8 two different ways that that are defined canonically equivalent and expecting the same result is empirically wrong

I assume you meant expecting _a different_ result.

As I mentioned [above](https://github.com/harfbuzz/harfbuzz/discussions/2942#discussioncomment-586115), sometimes "de facto" standards arise that contradict official standards. Uniscribe, by not normalizing, enables this nonstandard behavior prevalent in fonts for Biblical Hebrew. Typography may be like pop psychology, where "being an enabler" is not a good thing; it means, implicitly, being an enabler of some undesirable behavior. But nonetheless, these fonts, and their behavior, have been around for many years now.

Much of the de facto standard for the ways Unicode Biblical Hebrew fonts behave is, I think, due to John Hudson. Yet I think he is not at fault here: he is clear that the CGJ-optional behavior he encodes in SBL Hebrew is nonstandard. See the [SBL Hebrew Font User Manual](https://www.sbl-site.org/Fonts/SBLHebrewUserManual1.5x.pdf). Most hbo (Biblical Hebrew) fonts borrow (or at least started from) Hudson's logic for SBL Hebrew, and thus behave similarly to SBL Hebrew.

**Reply by alerque** [reply]
Yes, I did mean "a different result", sorry about the scrambled edit ;-)

**Reply by bdenckla** [reply]
I've now reviewed [UTR #53](http://unicode.org/reports/tr53/) more carefully, and I "grok" the issues better, though my understanding is still limited by my lack of Arabic.

A hypothetical HeMTRA could implement something close to Hudson's mark ordering, while still being something like AMTRA.

Such a HeMTRA could not be implemented, as AMTRA is, with an initial application of NFD or NFC, since that would destroy many orderings that were intentional in the input. Thus, while AMTRA "still works" even on normalized inputs, such a HeMTRA would not. So, such a HeMTRA might not be compatible with HarfBuzz. Does HarfBuzz apply AMTRA after normalization, or instead of normalization?

One aspect of Hudson's mark ordering could not be implemented by a HeMTRA whose output is, like AMTRA's, constrained to be canonically equivalent to its input.

That incompatible aspect is the introduction of a new "postpositive above" combining class 232. Since 3 of the four members of new class 232 come from old class 230, a HeMTRA introducing class 232 will give output that is canonically non-equivalent to some inputs. This is because the introduction of class 232 will cause some code points in old class 230 to be re-ordered relative to each other, a fundamental "no-no" if you are trying to preserve canonical equivalence.

I doubt that, in practice, any such inputs exist in the Biblical corpus. I.e. I doubt that, in practice, a HeMTRA introducing class 232 would in fact ever be given an input exposing that HeMTRA's ability to produce outputs canonically non-equivalent to its inputs. Such an input would be quite odd in that it would have a code point from new class 232 before a code point from new class 230. Because code points from new class 232 are postpositive, it is unclear what rendered appearance such an input would be intended to cause.

There are 5 words in the Hebrew Bible that require an above-accent (_telisha gedola_) to "shed" its _prepositive_ nature by appearing to the left of another above-accent, but I know of no analogous case where an above-accent is required to "shed" its _postpositive_ nature. Only a case like that would expose a Hudson-conforming HeMTRA to produce an output canonically non-equivalent to its input. Indeed this situation may explain part of why Hudson includes a "postpositive above" class but not a "prepositive above" class! Although _geresh muqdam_, AFAIK, could safely be the one and only member of a hypothetical "prepositive above" class, since _geresh muqdam_, unlike _telisha gedolah_, is never asked to "shed" its prepositive nature.

**Comment by tiroj** [comment]
Some background:

The Unicode canonical combining classes for Hebrew marks are widely recognised—including by past and present Unicode officers—as broken. My understanding is that they were originally provided by an Israeli standards body that had not adequately understood either the complexities of Biblical marking or the implications of the canonical combining classes for fonts.

I spent a lot of time with Microsoft engineers in the early 2000s getting their Hebrew shaping engine to work correctly with Biblical text, and one of the things that was necessary was to have a buffered display normalisation that would put the marks into an order that the font GSUB and GPOS features could consume, because the order coming out of Unicode normalisation was unworkable in a font format that has no adequate internal mechanism to reorder glyphs.

Some of the companies that were funding the SBL Hebrew font work also found that the Unicode Hebrew normalisation also wasn’t optimal for typical normalisation scenarios such as searching and sorting, so together we worked out the custom normalisation model that is an appendix to the [SBL Hebrew manual](https://www.sbl-site.org/Fonts/SBLHebrewUserManual1.5x.pdf). I believe this custom normalisation is pretty close to the buffered display normalisation that the Microsoft Hebrew engine performs. It is an intuitive ordering that works outwards from the base in groups of categories of interacting marks.

**Reply by bdenckla** [reply]
> the order coming out of Unicode normalisation was unworkable in a font format that has no adequate internal mechanism to reorder glyphs.

The GSUB mechanism is not adequate to "undo" Unicode normalization into a more font-friendly order?

I am just starting to learn about OpenType, so I may have an overly-optimistic sense of the power of GSUB, but I was hoping/imagining that, as annoying as it is to have to do so, GSUB could be used to:

1. First "undo" Unicode normalization into a more font-friendly order
2. Then proceed to do the "real work" of the font, e.g.
    1. substitute shin + shin_dot with shin_shin_dot
    2. substitute chataf_vowel + ZWJ + meteg with chataf_vowel_with_central_meteg, for each of the 3 chataf vowels
    3. etc.

**Comment by tiroj** [comment]
GSUB has no move operation, so the only way to perform reordering of glyphs within a run using GSUB is a series of contextual lookups that first add a duplicate of a glyph in a new position using a one-to-many substitution based on context, and then remove the original glyph from the run using a many-to-one substitution based on context. It is messy, and the more glyphs involved in context, the harder it is to do reliably. So while it is sometimes technically possible, it isn’t how OpenType Layout was designed to work. In OTL, standard reordering—which is essential for many scripts—is intended to be applied by the shaping engine, not by the font lookups. Obviously, for this to work in a predictable and compatible way across shaping engines, it needs to be accurately and publicly specified.

The frustrating thing about all this is that the preferred ordering for Hebrew marks is entirely intuitive if one knows anything at all about the layering of semantic information in Hebrew text: start with the consonant disambiguation (sing/shin) , then consonant modifiers (dagesh/ rafe), then vowel signs, then cantillation signs, working outwards from the base. Unicode’s canonical combining class assignments for Hebrew don’t follow this or any other sensible system: they are essentially arbitrary, so need to be overridden for display purposes and, I would argue, for better normalisation function too.

**Reply by bdenckla** [reply]
Thanks John, I see. So I take it neither SBL Hebrew, nor to your knowledge any other hbo font, attempts to "undo" Unicode normalization? I.e. SBL, and to your knowledge all hbo fonts, rely on the shaper to reorder for them?

I began to wonder how SBL & other hbo fonts look okay on platforms other than Windows, i.e. platforms with neither Uniscribe nor (its successor?) DirectWrite.

With great interest I discovered that HarfBuzz **does a subset of Hudson ordering**, documented and implemented in [hb-unicode.hh](https://github.com/harfbuzz/harfbuzz/blob/c39ab82c90479341dcf28eaa8174af6f08c0d7ae/src/hb-unicode.hh#L271).

This subset undoes the most glaring defects of Unicode NFD ordering by making sure the following marks appear before rather than after vowel marks: _shin_ dot, _sin_ dot, _dagesh_, & _rafeh_.

But it omits the more radical aspects of your recommended ordering. Notably, it omits the aspects of your ordering that would permit what I might call "Jerusalem without CGJ," which, not coincidentally, is the example that provoked me to launch this whole discussion!

There is also a different Hebrew reordering implemented in [hb-ot-shape-fallback.cc](https://github.com/harfbuzz/harfbuzz/blob/3fcf466b7b77f8eaf281f7a1a4e6ca90b5be54fe/src/hb-ot-shape-fallback.cc#L85); I'm not sure under what circumstances this fallback is resorted to.

**Reply by bdenckla** [reply]
Small update: as I learned from reading [Mozilla bug 662055](https://bugzilla.mozilla.org/show_bug.cgi?id=662055), HarfBuzz's permutation of Hebrew combining classes does permit a limit subset of "Jerusalem without CGJ" by letting "ah" vowels (_patach_ or _qamats_) stay before _hiriq_ or _sheva_. I am impressed that the parties in question were aware of (and addressed) the _sheva_ case,  since it only occurs in 4 words in the Bible! To my knowledge, no Biblical Hebrew (hbo) font handles the _sheva_ case quite right, i.e. no font gives _sheva_ the same special positioning that the _hiriq_ gets. So HarfBuzz is ahead of even the font experts on this particular score!

Unfortunately, the parties in question seem to have been unaware that this fix is limited to cases without trope. (For our purposes here at least, trope = cantillation = accents.) Even if they had been aware of this limitation, they would have probably accepted this limitation since removing this limitation requires a more radical reordering than can be accomplished with class permutation.

In other words, the HarfBuzz permutation fix applies only to what we might call "_siddur_ (prayerbook)" Hebrew, but not Biblical Hebrew. (Prayerbooks typically only have trope in a few small excerpts from the Hebrew Bible such as the _Ve-ahavta_.)

In Biblical Hebrew, i.e. in Hebrew with trope, "Jerusalem" always appears with a trope mark on the _lamed_, i.e. a trope mark between the two vowels of the _lamed_, except for one word in Psalm 147:12 where the trope for the word is _dehi_, which is a prepositive trope, and stress helpers are not used in the trope of Job, Proverbs, and Psalms, so the stressed syllable has no trope mark on it, i.e. the _lamed_ (the initial consonant for the stressed syllable) has no trope on it.

(Note that I refer to "the two vowels of the _lamed_," but the second vowel, _hiriq_ or _sheva_, only superficially "belongs" to the _lamed_, i.e. only "belongs" to the _lamed_ in a Unicode sense. Conceptually the second vowel belongs to the "phantom" _yod_ between the _lamed_ and the _mem_, and visually the goal is for it to appear either between the _lamed_ and the _mem_, or just barely underneath the right edge of the _mem_.)

**Reply by bdenckla** [reply]
In the [Unicode Normalization FAQ](https://www.unicode.org/faq/normalization.html), entries [9](https://www.unicode.org/faq/normalization.html#9) & [10](https://www.unicode.org/faq/normalization.html#10) touch on some of the issues we've been discussing here. Unfortunately these entries are sometimes wrong or at least misleading. Also, to me, their tone seems both dismissive and defensive.

Entry 9 reads:

> the rendering engine could reorder the marks internally in a different order if it wanted to, as long as that order was canonically equivalent. In particular, any permutation of the non-zero CCC values can be used for a canonically equivalent internal ordering. So internally a rendering engine could permute weights <27, 28,..., 32, 33> to <33, 27, 28,..., 32>, getting SHADDA before all vowel signs, for example. The restrictions are that only non-zero ccc values can be changed, and that they can only be permuted, not combined or split:
>
> * you can't reassign two characters that had the same ccc values to different values
> * you can't reassign two characters that had different ccc values to the same value
> * you can't change characters with ccc = 0.

I agree with entry 9's initial claim: for a re-ordering function's output to always be CE to its input, it is sufficient (but not necessary!) for the re-ordering function to be a stable sort based on a permutation of the nonzero combining classes.

But, then entry 9 goes on to claim (or seem to claim) that being a permutation is a **necessary** condition. In particular it claims that "you can't reassign two characters that had different ccc values to the same value." This is false. If the sort is stable, as, by definition, any sort using combining classes should be, you can combine (merge) (take the union of) two combining classes and still preserve CE. This is an important point since this is what we'd like to do for Hebrew: we'd like to merge the following:

- the class with only _meteg_ (22) 
- all 11 of the little 1- or 2-member below-vowel classes (classes 10-20)
- the big below-accent class (220).

In insisting (or seeming to insist) on permutations of combining classes, entry 9 also ignores the possibility of more complex re-ordering functions such as AMTRA. (AMTRA is not simply a stable sort based on a reassignment of combining classes.)

In forbidding the merging of combining classes, perhaps what this entry means to describe are re-ordering functions that are not just CE-preserving, but also robust to pre-normalization. I defined a function f to be "robust to pre-normalization" if and only if f(x) = f(g(x)) where g(x) is the normalized version of x, e.g. g(x) is x in NFD. This is the case for AMTRA, which indeed applies NFD as its first step!

Entry 10 claims:

> There was a problem [with Biblical Hebrew], but it has been addressed. Because the Hebrew points are defined to have distinct combining classes, their character semantics is such that their ordering is immaterial in the standard. To handle those cases where visual ordering is material, see the discussion of [CGJ] ...

This just asserts that the problem has been addressed without saying what the problem (singular?!) was or exactly how it was addressed. It seems to be saying the problem has been addressed by permutation of combining classes, but it does not document this permutation, nor does it provide a reference to such documentation.

It is true that the vast majority of problems with Biblical Hebrew combining classes can be solved by a permutation. But problems still remain, in the sense that, in contexts where there is a threat (or a certainty!) of normalization, CGJ is still needed to "defend" the order of marks in many cases that defy even the intuition of people who are experts in this writing system, if they are not experts in its unfortunate Unicode implementation:

1. In the Hebrew Bible, there are about 400 CGJs needed before _hiriq_ or _sheva_ when either is the last mark on a below-accented _lamed_ in words for Jerusalem.
2. In certain styles of Hebrew Bible, there are about 10 CGJs needed before _meteg_. While there are only about 10 of these, they are no "weirder" to experts than other cases not requiring CGJ. In other words, I believe the need for CGJ should not only be rare, it should be clear why it is needed in those cases. I assert that there are no such cases in Hebrew.

Reasonable people could disagree about what, if anything, to do about these remaining needs for CGJ that cannot be addressed by permutation. But it is unfortunate that the authors of this FAQ were either so dismissive of these issues as to not mention them or, (I hope this is the case) were unaware of these issues and therefore labored under the false impression that all issues related to Biblical Hebrew combining classes can be (and had been) solved by permutation.

**Comment by bdenckla** [comment, marked as answer]
My original question, also the title of this discussion, was the terse, interrobanged:
> harfbuzz normalizes?!

This could be expanded to: "HarfBuzz seems to normalize. Is that true? If so, why does HarfBuzz normalize, when Uniscribe/DirectWrite seems not to normalize?"

As to whether HarfBuzz normalizes, I can now definitively say, "sort of."  I.e. "sort of: it does a reordering based on a permutation of the canonical Unicode combining classes, implementing the most important parts of John Hudson's recommended mark ordering."

So on that basis alone I'll probably mark this comment as (the? an?) answer.

As to why HarfBuzz does this exactly this, I guess I'm still not 100% sure, but I can say that what HarfBuzz does strikes a reasonable compromise by implementing the most important parts of Hudson's recommendation while only using the rather simple and conservative mechanism of permutation. For example, only using permutation makes the reordering function robust to pre-normalization, like AMTRA is (see [above](https://github.com/harfbuzz/harfbuzz/discussions/2942#discussioncomment-600779) for a definition of "robust to pre-normalization").

My only remaining concerns/reservations are:

- This leaves HarfBuzz's behavior similar to, but still significantly different from Uniscribe/DirectWrite.
- HarfBuzz's compromise may have been made without "all the input"; in particular the discussion in [Mozilla bug 662055](https://bugzilla.mozilla.org/show_bug.cgi?id=662055), which seems to have motivated this work, ignores hundreds of Biblical locations that would benefit from a full (or fuller) implementation of Hudson's recommendation. Still, perhaps the HarfBuzz maintainers would have come to the same compromise, even with "all the input."

My main "take home" from all of this is an even greater commitment squashing all CGJ-related bugs in all important hbo fonts. Though web browsers' use of normalization on all platforms makes it very important that fonts are CGJ-compatible, HarfBuzz's use of normalization, which basically means Linux's use of normalization, makes CGJ compatibility even more crucial.

**Comment by behdad** [comment]
What we *are* open to do: improve rendering of sequences of Hebrew marks.

What we are *not* open to do: render two canonically-equivalent sequences differently.

Please open new issues if you want changes of the first kind and we'd happily consider.

Thanks.
